File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/ClarifyTool.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Text.Json

module ClarifyTool =

    let schema : ToolSchema =
        let toolName = ToolName.create "clarify_tool" |> Result.toOption |> Option.get
        {
            Name = toolName
            Description = "Ask the user one or more multiple-choice questions to resolve underspecified requirements or solicit design choices."
            ParametersJson = """{
                "type": "object",
                "properties": {
                    "question": { "type": "string", "description": "The question or design decision requiring user clarification." },
                    "options": { "type": "array", "items": { "type": "string" }, "description": "List of selectable options for the user." },
                    "is_multi_select": { "type": "boolean", "description": "Whether multiple options can be selected." }
                },
                "required": ["question", "options"]
            }"""
        }

    type ClarificationCallback = string -> string list -> bool -> Async<string>

    let createHandler (callback: ClarificationCallback option) : string -> Async<string> =
        fun (args: string) -> async {
            try
                use doc = JsonDocument.Parse(args)
                let root = doc.RootElement
                let question = root.GetProperty("question").GetString()
                let options =
                    root.GetProperty("options").EnumerateArray()
                    |> Seq.map (fun el -> el.GetString())
                    |> Seq.toList

                let isMultiSelect =
                    let mutable elem = JsonElement()
                    if root.TryGetProperty("is_multi_select", &elem) then
                        elem.GetBoolean()
                    else false

                match callback with
                | Some cb ->
                    let! userResponse = cb question options isMultiSelect
                    return sprintf "User selected: %s" userResponse
                | None ->
                    let defaultChoice = if options.Length > 0 then options.Head else "No option provided"
                    printfn "  [Clarify Tool] (Non-interactive mode) Defaulting to option: %s" defaultChoice
                    return sprintf "User selected (default): %s" defaultChoice
            with ex ->
                return sprintf "Error executing clarify_tool: %s" ex.Message
        }

    let register (registry: ToolRegistry) (callback: ClarificationCallback option) : unit =
        let handler = createHandler callback
        registry.Register("clarify_tool", schema.Description, handler, schema.ParametersJson)

File: implementations/csharp/src/Skight.AgentPlatform/ClarifyTool.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

namespace Skight.AgentPlatform
{
    public delegate Task<string> ClarificationCallback(string question, List<string> options, bool isMultiSelect);

    public static class ClarifyTool
    {
        public static readonly string SchemaJson = @"{
            ""type"": ""object"",
            ""properties"": {
                ""question"": { ""type"": ""string"", ""description"": ""The question or design decision requiring user clarification."" },
                ""options"": { ""type"": ""array"", ""items"": { ""type"": ""string"" }, ""description"": ""List of selectable options for the user."" },
                ""is_multi_select"": { ""type"": ""boolean"", ""description"": ""Whether multiple options can be selected."" }
            },
            ""required"": [""question"", ""options""]
        }";

        public static Func<string, Task<string>> CreateHandler(ClarificationCallback? callback = null)
        {
            return async (argsJson) =>
            {
                try
                {
                    using var doc = JsonDocument.Parse(argsJson);
                    var root = doc.RootElement;
                    string question = root.GetProperty("question").GetString() ?? "Clarification requested";
                    var options = root.GetProperty("options")
                        .EnumerateArray()
                        .Select(e => e.GetString() ?? string.Empty)
                        .Where(s => !string.IsNullOrEmpty(s))
                        .ToList();

                    bool isMultiSelect = root.TryGetProperty("is_multi_select", out var elem) && elem.GetBoolean();

                    if (callback != null)
                    {
                        string userResponse = await callback(question, options, isMultiSelect);
                        return $"User selected: {userResponse}";
                    }
                    else
                    {
                        string defaultChoice = options.Count > 0 ? options[0] : "No option provided";
                        Console.WriteLine($"  [Clarify Tool] (Non-interactive mode) Defaulting to option: {defaultChoice}");
                        return $"User selected (default): {defaultChoice}";
                    }
                }
                catch (Exception ex)
                {
                    return $"Error executing clarify_tool: {ex.Message}";
                }
            };
        }

        public static void Register(ToolRegistry registry, ClarificationCallback? callback = null)
        {
            registry.Register("clarify_tool", "Ask the user one or more multiple-choice questions to resolve underspecified requirements or solicit design choices.", CreateHandler(callback), SchemaJson);
        }
    }
}
